### PR TITLE
feat(typescript-estree): skip `isTTY` version check if user passes `loggerFn` on unsupported TypeScript version warning

### LINKED ISSUE #3506: Skip the `isTTY` check for printing the "SUPPORTED TYPESCRIPT VERSIONS" message if the user passes in the `loggerFn` option
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I spent 30mins tonight trying to determine why this warning was being displayed sometimes, but then it would disappear whenever I tried to debug it:

> WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

Is it a caching problem?  Is the TypeScript version actually correct when invoking ESLint programmatically?  No, turns out the magic behavior is based on `process.stdout?.isTTY`:

https://github.com/typescript-eslint/typescript-eslint/blob/44452dd5a545665c2365f2f747fc7610ea0eb686/packages/typescript-estree/src/parser.ts#L323-L330

Maybe the intent was to hide the warning from a CI build, but show it when humans invoke ESLint? That was not the effect. The effect was nondeterministic behavior that was frustrating to investigate.  Arguably this is notifying about a nontrivial problem; we don't really want our production builds linting with an outdated TypeScript parser.

Would you accept a PR to delete this `process.stdout?.isTTY` logic and always show the warning regardless of how ESLint is invoked?

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `4.26.1` |
| `TypeScript`                           | `4.3.2` |
| `node`                                 | `12.20.1` |

### PR DESCRIPTION
**Pull request by hyperupcall**
## PR Checklist

- [x] Addresses an existing open issue: fixes #3506
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

From the linked PR, this disabled the `isTTY` heuristic if `loggerFn` was supplied as an option:
- If `loggerFn` is a function, then the heuristic is now bypassed completely
- If `loggerFn` is false, then the heuristic logic runs, but the result is effectively a no-op (nothing is printed in this case)

Additionally, this fixes `warn-on-unsupported-ts.test.ts` by properly requiring/importing `parser.ts` for every test. Previously, it would only import once at the start, which prevented the top-level assignment `let warnedAboutTSVersion = false` (and later assignments/side-effects thereof) from being reset (in `warnAboutTSVersion.ts`).

It seems TypeScript complains about the `await import`; the error seems ignored [in other places](https://github.com/typescript-eslint/typescript-eslint/blob/e50d1a39a057b006b99bb47718122728b319eb92/packages/eslint-plugin/tools/generate-configs.ts#L32C43-L32C43) so I ignored it too.

### CODE REVIEW

File: packages/typescript-estree/src/parseSettings/warnAboutTSVersion.ts [hunk 1]
```diff
@@ -23,23 +23,32 @@ const isRunningSupportedTypeScriptVersion = semver.satisfies(
...
+
+    if (passedLoggerFn) {
+      parseSettings.log(versionWarning);
+    } else {
+      const isTTY =
+        typeof process === 'undefined' ? false : process.stdout?.isTTY;
+      if (isTTY) {
+        parseSettings.log(versionWarning);
+      }
     }
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/parseSettings/warnAboutTSVersion.ts:50]: [Cleanup] Nitpicking: could reduce a bit of code here and not initialize the variables if we don't need, thoughts?

```suggestion
    if (
      passedLoggerFn ||
      (typeof process !== 'undefined' && process.stdout?.isTTY)
    ) {
      const border = // ...
      const versionWarning = // ...
      parseSettings.log(versionWarning);
    }
```

I also kind of like reducing nesting of the function by starting off with:

```ts
if (isRunningSupportedTypeScriptVersion || warnedAboutTSVersion) {
  return;
}
```

(this is all nitpicking, no worries if you hate it and don't want it)

File: packages/typescript-estree/tests/lib/warn-on-unsupported-ts.test.ts [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
 jest.mock('semver');
 
 const resetIsTTY = process.stdout.isTTY;
 
 describe('Warn on unsupported TypeScript version', () => {
+  let parser: typeof Parser;
+
+  beforeEach(async () => {
+    // @ts-expect-error -- We don't support ESM imports of local code yet.
+    parser = await import('../../src/parser.ts');
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/lib/warn-on-unsupported-ts.test.ts:14]: [Cleanup] Nit: we don't generally use the `.ts` extensions, even if they are supported in the testing framework. No need.

```suggestion
    parser = await import('../../src/parser');
```

Confirming that trying to add the explicit `.js` in gives the unfortunate Jest complaint that the file can't be found.

**Review by JoshuaKGoldberg — APPROVED** [review]
💯 great, thanks! Nice to see this 3+ year old issue resolved.

Just a few nitpicks from my end. Because they're not blocking at all I can just touch them up before merging.

Cartoon pickle smiling and lifting a top hat up and down. Caption: 'Thanks!'

**Comment by hyperupcall** [comment]
> Just a few nitpicks from my end. Because they're not blocking at all I can just touch them up before merging.

Sounds good - I'm not sure if my branch is comitable from your end, so feel free to point out the nits, and I'll be able to make the changes if you are not able to

**Comment by JoshuaKGoldberg** [comment]
Ah you're right! It's not. If you could make them or grant me access, that'd be great thanks 😄. Otherwise no worries.

File: packages/typescript-estree/src/parseSettings/warnAboutTSVersion.ts (hunk 1 above)
Comment (hyperupcall) [reply packages/typescript-estree/src/parseSettings/warnAboutTSVersion.ts]: I applied your suggestions! I think they read better

**Review by JoshuaKGoldberg — APPROVED** [review]

**Pull request merged**